File: code/naive-bayes-classifier/python/training_set.py
from collections import defaultdict

training_set = defaultdict(list)


def add_document(message: str, cat: str):
    training_set[cat].append(message)


add_document('Hello viagra', 'spam')
add_document('viagra again', 'spam')
add_document('viagra strikes', 'spam')
add_document('Hello world', 'notSpam')

File: code/naive-bayes-classifier/rust/src/training_set.rs
use std::collections::hash_map;

#[derive(Debug)]
pub struct TrainingSet {
    pub store: hash_map::HashMap<String, Vec<String>>
}

impl TrainingSet {
    pub fn new() -> Self {
        TrainingSet {
            store: hash_map::HashMap::new()
        }
    }

    pub fn add_document<T: Into<String>>(&mut self, msg: T, cat: T) {
        self.store.entry(cat.into()).or_default().push(msg.into());
    }

    pub fn labels(&self) -> impl Iterator<Item=&String> {
        self.store.keys()
    }
}

pub fn init() -> TrainingSet {
    let mut training_set = TrainingSet::new();

    training_set.add_document("Hello viagra", "spam");
    training_set.add_document("viagra again", "spam");
    training_set.add_document("viagra strikes", "spam");
    training_set.add_document("Hello world", "notSpam");

    training_set
}
